src/poliastro/core/iod.py

```diff
@@ -111,42 +111,75 @@ def vallado(k, r0, r, tof, short, numiter, rtol):
 
     A = t_m * (norm_r * norm_r0 * (1 + cos_dnu)) ** 0.5
 
-    if A == 0.0:
+    if abs(A) <= 0.2:  # Not exactly zero
         raise RuntimeError("Cannot compute orbit, phase angle is 180 degrees")
 
-    psi = 0.0
-    psi_low = -4 * np.pi ** 2
-    psi_up = 4 * np.pi ** 2
+    psi = 0
+
+    if M == 0:  # Usual M == 0 case
+        psi_low = -4.0 * np.pi ** 2
+        psi_up = 4.0 * np.pi ** 2
+
+    else:
+        psi_low = 4.0 * M ** 2 * np.pi ** 2
+        psi_up = 4.0 * (M + 1.0) ** 2 * np.pi ** 2
+
+    psi = _initial_guess_vallado(psi, psi_low, psi_up, M, tof, direct)
 
     count = 0
+    ynegktr = 1
+    tof_new = -10
+
+    while count < numiter and ynegktr <= 10:
+
+        if abs(c2(psi) > rtol):
+            y = norm_r0_plus_norm_r + A * (1.0 - psi * c3(psi)) / c2(psi) ** 0.5
+        else:
+            y = norm_r0_plus_norm_r
 
-    while count < numiter:
-        y = norm_r0_plus_norm_r + A * (psi * c3(psi) - 1) / c2(psi) ** 0.5
         if A > 0.0:
             # Readjust xi_low until y > 0.0
             # Translated directly from Vallado
-            while y < 0.0:
-                psi_low = psi
+            ynegktr = 1
```

Comment:
Can we rename this lovely variable `ynegktr` to `neg_y_max_iter`?

Reply:
Too long :( . Anything better ?

Reply:
This is not FORTRAN, we can afford `neg_y_max_iter` 😝

Reply:
Haha, okay!